src/stats/probs.rs

```diff
@@ -0,0 +1,352 @@
+// Copyright 2014-2016 Johannes Köster.
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Handling log-probabilities.
+
+use std::mem;
+use std::f64;
+use std::iter;
+use std::ops::{Add, Sub, Mul, Div};
+use num::traits::cast;
+use num::NumCast;
+
+use itertools::linspace;
+use itertools::Itertools;
+use itertools::misc::ToFloat;
+
+
+/// A factor to convert log-probabilities to PHRED-scale (phred = p * `LOG_TO_PHRED_FACTOR`).
+const LOG_TO_PHRED_FACTOR: f64 = -4.3429448190325175; // -10 * 1 / ln(10)
+
+
+/// A factor to convert PHRED-scale to log-probabilities (p = phred * `PHRED_TO_LOG_FACTOR`).
+const PHRED_TO_LOG_FACTOR: f64 = -0.23025850929940456; // 1 / (-10 * log10(e))
+
+
+/// Calculate log(1 - p) with p given in log space without loss of precision as described in
+/// http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf.
+fn ln_1m_exp(p: f64) -> f64 {
+    assert!(p <= 0.0);
+    if p < -0.693 {
+        (-p.exp()).ln_1p()
+    } else {
+        (-p.exp_m1()).ln()
+    }
+}
+
+custom_derive! {
+    #[derive(
+        NewtypeFrom,
+        NewtypeDeref,
+        NewtypeAdd(*),
+        NewtypeSub(*),
+        NewtypeMul(*),
+        NewtypeDiv(*),
+        PartialEq,
+        PartialOrd,
+        Copy,
+        Clone,
+        Debug,
+        Default
+    )]
+    /// A newtype for probabilities.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// #[macro_use]
+    /// extern crate approx;
+    /// # extern crate bio;
+    /// # fn main() {
+    /// use bio::stats::Prob;
+    ///
+    /// let p = Prob(0.5);
+    /// let q = Prob(0.2);
+    ///
+    /// assert_relative_eq!(*(p + q), *Prob(0.7));
+    /// # }
+    /// ```
+    pub struct Prob(pub f64);
+}
+
+
+custom_derive! {
+    #[derive(
+        NewtypeFrom,
+        NewtypeDeref,
+        NewtypeAdd(*),
+        NewtypeSub(*),
+        PartialEq,
+        PartialOrd,
+        Copy,
+        Clone,
+        Debug
+    )]
```

Comment:
Nit: is it perhaps preferable to have the derive block below the block comment? Or is that incompatible with the way custom_derive works?